feat!: UP-1130 teleadr fixes

baronbosse opened the pull request:
This handles multiple schemas in the same WSDL file better than before.

Also there are some breaking changes in `ews:decode_in` to handle deserialization  of  endpoints with headers and the possiblity of multiple messages in the same SOAP call.

baronbosse committed:
fix: temporarily remove options that don't work

baronbosse committed:
fix: handle more than one schema in the same WSDL file

baronbosse committed:
fix: handle more than one SOAP version port definition

baronbosse committed:
fix: use correct type and use="required" type

baronbosse committed:
fix: put all types with refs in retry pile

kivra-kenand committed:
fix: encode Value with BaseType

baronbosse committed:
fix: don't delete atrributes

baronbosse committed:
chore: add failing test for split attributes

baronbosse committed:
fix: don't split attributes with space

baronbosse committed:
fix: handle multiple schemas in one file better for local

baronbosse committed:
fix: bug in split_schemas

baronbosse committed:
feat: add failing testcase for int attrs

baronbosse committed:
feat: handle decoding with headers

baronbosse committed:
fix: decode attributes to correct erlang type

baronbosse committed:
feat: don't output undefined when attrs are required

baronbosse committed:
feat: test everything in an wsdl

baronbosse committed:
chore: switch to logger

baronbosse committed:
fix: ignore a ews bug for now

baronbosse committed:
fix: test enums

baronbosse committed:
fix: handle list of enums

baronbosse committed:
chore: document more breaking changes

baronbosse commented on src/ews_xml.erl:
```diff
@@ -84,7 +84,7 @@ emit_tag({{Ns, Name}, Attributes, Children}, Nss) ->
     QName = {Prefix, Name},
     [emit_start_tag(QName),
-     emit_attributes(lists:ukeysort(2, XmlNsDecl++Attributes), NewNss),
+     emit_attributes(lists:keysort(2, XmlNsDecl++Attributes), NewNss),
```
This is just wrong. The change fixes a bug.

baronbosse committed:
feat: select only soap11 bindings and ports for now

This is a bit more robust that just selecting the first port.

kivra-kenand commented on src/ews_xsd.erl:
```diff
@@ -586,11 +630,19 @@ process([#element{name=Qname, type=undefined, parts=Ps} = E | Rest], Retry, Ts,
-            ews_model:put(Type, Model, TypeMap),
-            process(Rest, Retry2, Ts, [Type | AccWithSubTypes], [Elem | ElemAcc],
-                    TypeMap, Model, Parent, AttrAcc);
+            case [ Ref || #element{type=#reference{}} = Ref <- SubElems] of
```
Could be helpful with a helpful comment here :D

baronbosse replied:
done

kivra-kenand commented on src/ews_xsd.erl:
```diff
@@ -55,44 +58,54 @@
-    PrSchemas = [ #schema{namespace=Ns,
-                          url=Url,
-                          types=parse_types(S)} || {Ns, Url, S} <- Schemas ],
+parse_schema(Schemas0, {Acc, Model}) when is_atom(Model) ->
```
please specs on apis

kivra-kenand commented on src/ews_wsdl.erl:
```diff
@@ -252,15 +269,14 @@ parse_types(WsdlDoc, Model) ->
     Schemas = wh:get_children(Types, "schema"),
     {Res, _} =
-        lists:foldl(fun ews_xsd:parse_schema/2, {undefined, Model}, Schemas),
+        ews_xsd:parse_schema(Schemas, {undefined, Model}),
```
see if the 2nd argument in the response tuple is needed

kivra-kenand commented on src/ews_xsd.erl:
```diff
@@ -55,44 +58,54 @@
 do_get_all_schemas({Ns, Base, Schema}, Acc) ->
     case find_imports(Schema) of
         [] ->
             [{Ns, Base, Schema} | Acc];
```
perhaps Ns can be dropped in this tuple?

baronbosse replied:
I tried to remove it, and then a lot of bugs started happening, references could be resolved and so forth.

So it is too scary to remove.

kivra-kenand commented on src/ews_xml.erl:
```diff
@@ -198,18 +198,45 @@ to_txt(TxtLst) ->
+outside_quote([C | T], [_|_] = Acc, Attrs) when
+      C == $ ; C == $\t; C == $\r; C == $\n ->
+    outside_quote(T, [], [lists:reverse(Acc) | Attrs]);
+outside_quote([C | T], [], Attrs)when
```
nit: missing whitespace

baronbosse replied:
done

kivra-kenand commented on src/ews_xml.erl:
```diff
@@ -198,18 +198,45 @@ to_txt(TxtLst) ->
+    inside_quote(T, [C | Acc], Attrs);
+outside_quote([C | T], Acc, Attrs) ->
+    outside_quote(T, [C | Acc], Attrs);
+outside_quote([], Acc, Attrs) ->
```
```
outside_quote([], [], Attrs) ->
  Attrs;
```
Perhaps add more unit-tests?

kivra-kenand replied:
example: `<foo >` would fail now we think

baronbosse replied:
fixed

kivra-kenand commented on src/ews_xml.erl:
```diff
@@ -84,7 +84,7 @@ emit_tag({{Ns, Name}, Attributes, Children}, Nss) ->
     QName = {Prefix, Name},
     [emit_start_tag(QName),
-     emit_attributes(lists:ukeysort(2, XmlNsDecl++Attributes), NewNss),
+     emit_attributes(lists:keysort(2, XmlNsDecl++Attributes), NewNss),
```
perhaps add a comment why ukeysort should not be used here

kivra-kenand commented on src/ews_test.erl:
```diff
@@ -0,0 +1,147 @@
+    {ok, Info} = ews:get_service_op_info(Model, Service, Op),
+    #{in_hdr := InHdr, in := In, out_hdr := OutHdr, out := Out,
+      fault := Fault} = maps:from_list(Info),
+    {ok, FaultInfo} = ews_svc:get_op_message_details(Model, Service, Op),
```
this could be moved into test_fault

kivra-kenand commented on src/ews_test.erl:
```diff
@@ -0,0 +1,147 @@
+    FaultVals = def_vals(Fault, Tbl),
+    logger:notice("FaultVals: ~tp~n", [FaultVals]),
+    FaultSOAP = ews:encode_service_op_faults(Model, Service, Op,
+                                             <<"9000">>, <<"boo">>, FaultVals),
```
Could you attibute <<"9000">> and <<"boo">> to descriptive variables

kivra-kenand commented on src/ews_test.erl:
```diff
@@ -0,0 +1,147 @@
+-module(ews_test).
```
document what this is for and add to README as to why this is in src

baronbosse replied:
done

kivra-kenand commented on src/ews_svc.erl:
```diff
@@ -200,10 +201,9 @@ decode_in(ModelRef, SOAP) ->
+                {ok, Hdrs, Res} ->
+                    {ok, {Svc, OpName, Hdrs, Res}}%%;
                 %% {error, Reason} ->
                 %%     {error, {Reason, {Svc, OpName, Op}}}
```
consider removing or re-adding

kivra-kenand commented on src/ews_svc.erl:
```diff
@@ -774,43 +800,53 @@ try_encode_fault(Part, [#elem{} = Fault | Faults], Model) ->
     OutHdrs = proplists:get_value(out_hdr, Info),
     DecodedHeaders = decode_headers(Headers, OutHdrs, Model, Opts),
-    [DecodedBody] = ews_serialize:decode(Body, Outs, Model),
+    DecodedBody = ews_serialize:decode(Body, Outs, Model),
```
s/DecodedBody/DecodedBodies/g

baronbosse committed:
chore: comment about child references

baronbosse committed:
chore: remove unused Accs

The code no longer folds over the schemas in the same way.

baronbosse committed:
fix: don't sort the attributes, they are in the order they should be

baronbosse committed:
chore: comment on the use of keysort

baronbosse committed:
chore: whitespace

baronbosse committed:
fix: add failing test for trailing whitespace in tag

baronbosse committed:
fix: don't add empty list to args on trailing whitespace in tag

baronbosse committed:
chore: remove unneaded code

baronbosse committed:
chore: rename Body to Bodies

baronbosse committed:
chore: change copyright dates

baronbosse committed:
chore: comment on how to use ews_test

baronbosse committed:
chore: move function call

baronbosse committed:
chore: give more descriptive names to code and string

baronbosse committed:
chore: add readme section about ews_test

baronbosse committed:
feat: sort types to be consistent between otp versions and OSes

This will probably re-sort you generated deps once, but going forward
there will be less moving around of types in the generated .hrl-files.

baronbosse committed:
chore: spec up api
